hazelcast/src/main/java/com/hazelcast/internal/util/counters/SwCounter.java

```diff
@@ -22,24 +22,45 @@
 /**
  * A {@link Counter} that is made to be used by a single writing thread.
  * <p>
- * It makes use of the opaque read/writes to provide a lower overhead than a volatile write. Volatile write is
- * pretty expensive. On the X86 is causes subsequent loads to wait till the store in the store buffer has been
- * written to the coherent cache. And this can take some time because it could be that a whole bunch of cache
- * lines need to be invalidated and this can add a lot of latency to those loads. Opaque doesn't provide any
- * ordering guarantees with respect to other variables. It is atomic and coherent and it is super well suited for
- * progress indicators like performance counters. Opaque is primary
+ * It makes use of the opaque read/writes to provide a lower overhead than a
+ * volatile read/write. Volatile reads/writes can be pretty expensive. If a
+ * volatile write is followed by a volatile read to a different variable, then
+ * on the X86 is causes that read and all subsequent reads to to stall till the
+ * stores in the store buffer has been written to the coherent cache. And this can
+ * take some time because it could be that there one or more stores in the store buffer
+ * (on Skylake the store buffer can contain 50+ stores) each each of these stores
+ * needs to wait for the cache line to be successfully invalidated on the other CPUs.
+ * And this can add a lot of latency to those loads and any instruction depending
+ * on the loaded values; so you can end up with a core idling because few instructions
+ * are ready for execution. Since there are roughly 10 LFBs (Line Fill Buffers)
+ * on modern Intel processor at most 10 cache lines can be invalidated in parallel.
+ * <p/>
+ * Unlike volatile, opaque doesn't provide any ordering guarantees with respect
+ * to other variables. So it will not lead to loads stalling till the store buffer
+ * has been drained.
+ * <p/>
+ * Opaque does provide:
+ * <ol>
+ *     <li>atomicity: so no torn reads/writes</li>
+ *     <li>coherence: (1) you don't go back reading an older version after you read a
+ *     newer values and (2) cores will not disagree upon the order of writes to a
+ *     single variable)
+ *     </li>
+ * </ol>
+ * And therefor is super well suited for progress indicators like performance
+ * counters.
  * <p>
  * This counter does not provide padding to prevent false sharing.
  * <p>
- * One might wonder why not use the AtomicLong.inc. The problem here is that AtomicLong requires a full fence,
- * so there is waiting for store and load buffers to be drained. This is more expensive.
- * <p>
- * One might also wonder why not use the following:
- * <pre>
- *     atomicLong.lazySet(atomicLong.get()+1)
- * </pre>
- * This causes a lot of syntactic noise due to lack of abstraction.
- * A counter.inc() gives a better clue what the intent is.
+ * I believe the design of the Counter object is a bit stale. The problem is
```

Comment:
I think the style of this paragraph can be updated so it doesn't express "belief" but rather explains a use case ("When several metrics fields need to be monitored") and proposes the better solution ("make a single object with multiple primitive fields ..."). Adding a `@see` pointing to example code where this is applied from tpc engine module would be great, instead of the rather generic last statement.

Reply:
Fixed